### ISSUE TITLE
How to get gamepad events after 1.12 FlutterView deprecation

### ISSUE THREAD

**Issue by Mithrandir2k18** [flutter/flutter#50789]
[Github issue on plugin using deprecated FlutterView](https://github.com/RainwayApp/flutter_gamepad/issues/3#issue-558180218)
[Related question on StackOverflow](https://stackoverflow.com/questions/60152919/how-to-exchange-androidtouchkeyprocessor-after-1-12-updateflutterview-deprecat)

This issue is probably only lack of documentation; still an issue. I cannot figure out how to upgrade the Android{Touch|Key}Processor, which was previously done by getting them via the FlutterView API, at least in the linked Gamepad Plugin.

## Steps to Reproduce
Try to compile a flutter app in android using the [flutter gamepad plugin](https://pub.dev/packages/flutter_gamepad).
```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.example.acu_remote_control, PID: 15345
    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.example.acu_remote_control/com.example.acu_remote_control.MainActivity}: java.lang.UnsupportedOperationException: The new embedding does not support the old FlutterView.
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3270)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3409)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.UnsupportedOperationException: The new embedding does not support the old FlutterView.
        at io.flutter.embedding.engine.plugins.shim.ShimRegistrar.view(ShimRegistrar.java:82)
        at com.example.flutter_gamepad.FlutterGamepadPlugin$Companion.registerWith(FlutterGamepadPlugin.kt:61)
        at com.example.flutter_gamepad.FlutterGamepadPlugin.registerWith(Unknown Source:2)
        at io.flutter.plugins.GeneratedPluginRegistrant.registerWith(GeneratedPluginRegistrant.java:18)
        at com.example.acu_remote_control.MainActivity.configureFlutterEngine(MainActivity.kt:10)
        at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onAttach(FlutterActivityAndFragmentDelegate.java:185)
        at io.flutter.embedding.android.FlutterActivity.onCreate(FlutterActivity.java:418)
        at android.app.Activity.performCreate(Activity.java:7802)
        at android.app.Activity.performCreate(Activity.java:7791)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1299)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3245)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3409) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7356) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930) 
I/Process: Sending signal. PID: 15345 SIG: 9
Process 15345 terminated.
```

How can we extend the key and touch processor functionality to handle [key and motion events](https://developer.android.com/training/game-controllers/controller-input.html) generated by a gamepad?

**Comment by lynn** [comment]
Hi, I'm the author of the `flutter_gamepad` plugin referenced in this issue.

This seems related to https://github.com/flutter/flutter/issues/40423.

Maybe @xster, who responded with a solution there, has an idea for how to tackle this use of FlutterView.

The way we rely on FlutterView being available is a little evil. We use reflection to swap in a "patched" AndroidKeyProcessor and AndroidTouchProcessor that supports gamepad buttons and axes:

```kt
val view: FlutterView = registrar.view()

fun viewField(name: String): Field {
   val field = FlutterView::class.java.getDeclaredField(name)
   field.isAccessible = true
   return field
}

// Hack: swap in a new AndroidTouchProcessor.
val touchProcessorField = viewField("androidTouchProcessor")
val rendererField = viewField("flutterRenderer")
val renderer = rendererField.get(view) as FlutterRenderer
val touchProcessor = GamepadAndroidTouchProcessor(renderer)
touchProcessorField.set(view, touchProcessor)

// Hack: swap in a new AndroidKeyProcessor.
val keyProcessorField = viewField("androidKeyProcessor")
val keyEventChannelField = viewField("keyEventChannel")
val textInputPluginField = viewField("mTextInputPlugin")
val keyEventChannel = keyEventChannelField.get(view) as KeyEventChannel
val textInputPlugin = textInputPluginField.get(view) as TextInputPlugin
val keyProcessor = GamepadAndroidKeyProcessor(keyEventChannel, textInputPlugin)
keyProcessorField.set(view, keyProcessor)
```

Maybe this is a sign that our plugin is simply trying to do more than a plugin should, and this should be patched into Flutter itself. (Actually, listening to gamepad _buttons_ is already possible, because they are simple key events — but not sticks and analog triggers.) Anyway, it would be nice to have a stopgap solution.

**Comment by xster** [comment]
I may or may not have understood the code snippet. What do your Gamepad* classes do exactly? Relating to the general sentiment (of accessing FlutterViews in the new plugins APIs), it's open for debate. We just haven't found a proper valid case so far where a specific `FlutterView` instance is needed in a plugin rather than a generic more general `android.view.View`. 

If some keypresses that need to be available in Dart is not, we can modify the engine itself as well to make sure events are not missed.

**Comment by lynn** [comment]
@xster [GamepadAndroidKeyProcessor](https://github.com/RainwayApp/flutter_gamepad/blob/948179994895673b221d5ec5433390925a3bd87b/android/src/main/kotlin/com/example/flutter_gamepad/FlutterGamepadPlugin.kt#L22-L52) is a subclass of FlutterView's AndroidKeyProcessor, that overrides *onKeyDown* and *onKeyUp* to basically do the following:

* If the incoming KeyEvent is a gamepad button press (from e.g. a Bluetooth Xbox controller), intercept it and turn it into a message on the plugin's event channel.
* Otherwise, call `super` and let FlutterView / AndroidKeyProcessor handle the event normally.

Similarly, GamepadAndroidTouchProcessor is a subclass of AndroidTouchProcessor that overrides *onGenericMotionEvent* to intercept joystick motion and turn it into platform event channel messages, but passes all other motion events through to Flutter.

(FWIW, I believe that right now, we get all the gamepad buttons as keypresses, but joystick motion isn't available in Dart at all. This plugin intercepts gamepad buttons anyway, just so that all gamepad input can conveniently be handled in the same way / at the same point.)

**Comment by xster** [comment]
I may or may not be answering the question you're looking for: you probably don't need the AndroidKeyProcessor which is just a Flutter implementation detail on how to handle keys. You can still just use the general Android APIs to catch gamepad keys. For instance, implement the new ActivityAware interface in your v2 plugin. When there's an activity, get the generic root android.view.View. Set an OnKeyListener on it. 

Maybe the problem is that the AndroidKeyProcessor on the dispatchKeyEvent blocked subsequent OnKeyListeners?

**Comment by xster** [comment]
Also @gspencergoog might have some opinions since he was just working on those key handling parts.

**Comment by aevzp** [comment]
Any update, guys?

**Comment by gspencergoog** [comment]
You should be able to use the generic Android APIs as @xster suggests. We don't block subsequent listeners, as far as I know.

**Comment by aevzp** [comment]
Now I'm trying to solve the problem of getting data from the controller sticks.

From controller we can get all button event using `RawKeyboardListener`. It's good, but don't enough.
I tried to do the following:

In Android I created `View` like an documentation: https://developer.android.com/training/game-controllers/controller-input#joystick
```
class GamepadView extends View {
    public GamepadView(Context context) {
        super(context);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        Log.e("TOUCH", event.toString());
        return super.onTouchEvent(event);
    }

    @Override
    public boolean onGenericMotionEvent(MotionEvent event) {
        Log.e("MOTION", event.toString());
        return super.onGenericMotionEvent(event);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        Log.e("onKeyDown", event.toString());
        return super.onKeyDown(keyCode, event);
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        Log.e("onKeyUp", event.toString());
        return super.onKeyUp(keyCode, event);
    }
}
```

Then following the guide
https://flutter.dev/docs/development/platform-integration/platform-views?tab=android-platform-views-java-tab#on-the-platform-side 
I transferred this View to Flutter's application.

GamepadControllerPlugin.java
```
public class GamepadControllerPlugin implements FlutterPlugin, MethodCallHandler {
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
      channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "gamepad_controller");
      channel.setMethodCallHandler(this);
  
      flutterPluginBinding
              .getPlatformViewRegistry()
              .registerViewFactory("hybrid-view-type", new NativeViewFactory());
  }
}
```

NativeView.java
```
class NativeView implements PlatformView {
    @NonNull
    private final GamepadView gamepadView;

    NativeView(@NonNull Context context, int id, @Nullable Map<String, Object> creationParams) {
        gamepadView = new GamepadView(context);
        gamepadView.setBackgroundColor(Color.parseColor("#252525"));
    }

    @NonNull
    @Override
    public View getView() {
        return gamepadView;
    }

    @Override
    public void dispose() {}
}
```

NativeViewFactory.java
```
public class NativeViewFactory extends PlatformViewFactory {
    public NativeViewFactory() {
        super(StandardMessageCodec.INSTANCE);
    }

    @NonNull
    @Override
    public PlatformView create(Context context, int id, Object args) {
        final Map<String, Object> creationParams = (Map<String, Object>) args;
        return new NativeView(context, id, creationParams);
    }
}
```

And then in the Flutter I added this `View`:
main.dart
```
  Widget build(BuildContext context) {
    return PlatformViewLink(
      viewType: 'hybrid-view-type',
      surfaceFactory:
          (BuildContext context, PlatformViewController controller) {
            return AndroidViewSurface(
              controller: controller,
              gestureRecognizers: const <Factory<OneSequenceGestureRecognizer>>{},
              hitTestBehavior: PlatformViewHitTestBehavior.opaque,
            );
      },
      onCreatePlatformView: (PlatformViewCreationParams params) {
        return PlatformViewsService.initSurfaceAndroidView(
          id: params.id,
          viewType: 'hybrid-view-type',
          layoutDirection: TextDirection.ltr,
          creationParams: creationParams,
          creationParamsCodec: StandardMessageCodec(),
        )
          ..addOnPlatformViewCreatedListener(params.onPlatformViewCreated)
          ..create();
      },
    );
  }
```

-------
As you can see, I added `Log.e` methods in `Gamepad` class to display caught events.
When I start testing this event `onKeyDown`, `onKeyUp` and `onGenericMotionEvent` never fired.
Event `onTouch` work fine. Also event `onGenericMotionEvent` fired only if I use sensor panel on the top on DualShock + if I press to sensor panel `Click` event is fired.
![IMG_6185](https://user-images.githubusercontent.com/52816901/103743645-30e05980-5005-11eb-8e1f-d7e19813ce52.jpg)

Attempting to use controller sticks does not trigger any native events.
-------
I also tried to do as @xster said. For my plugin I added new ActivityAware interface, get android.view.View and set listener `OnKeyListener`. Unfortunately it didn't work.

For now I totally stuck. Getting all buttons event from controller - it's good, but controller stick it's important part of gamepad =(

**Comment by aevzp** [comment]
I managed to listen to events from the controller sticks.

For this I created native `EditText` class:
```
class GamepadView extends EditText {
    private EventChannel.EventSink mEventSink;

    public GamepadView(Context context, EventChannel.EventSink eventSink) {
        super(context);
        this.mEventSink = eventSink;
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        // Without focus keys doesn't catch in Flutter
        this.requestFocus();
        super.onLayout(changed, left, top, right, bottom);
    }

    // Below in the code are the listeners: onGenericMotionEvent. onTouchEvent, onKeyUp
}
```

This view following the [documentation](https://flutter.dev/docs/development/platform-integration/platform-views?tab=android-platform-views-java-tab#performance) I am throwing into flutter.
There are certain limitations that I have researched, but this is not the point.

In `GamepadControllerPlugin.java` I have register this view
```
mFlutterPluginBinding
                  .getPlatformViewRegistry()
                  .registerViewFactory(renderName, new NativeViewFactory(events));
```

And then displayed it in Flutter inside method build:
```
    return Container(
      height: 100,
      child: PlatformViewLink(
        viewType: viewName,
        surfaceFactory: (BuildContext context, PlatformViewController controller) {
          return AndroidViewSurface(
            controller: controller,
            gestureRecognizers: const <Factory<OneSequenceGestureRecognizer>>{},
            hitTestBehavior: PlatformViewHitTestBehavior.opaque,
          );
        },
        onCreatePlatformView: (PlatformViewCreationParams params) {
          var service = PlatformViewsService.initSurfaceAndroidView(
            id: params.id,
            viewType: viewName,
            layoutDirection: TextDirection.ltr,
            creationParams: creationParams,
            creationParamsCodec: StandardMessageCodec(),
          )
            ..addOnPlatformViewCreatedListener(params.onPlatformViewCreated)
            ..create();

          return service;
        },
      ),
    );
```
In my case, focusing on the native input field does not interfere with my task.

But at the same time on the needed screen, I also use `Texture` [Widget](https://api.flutter.dev/flutter/widgets/Texture-class.html). And when I use displaying native input field for listen controller sticks and `Texture` widget, `Texture` widget is freeze and I have a lot of message in console like this:
```
....
D/DecorView(27057): semSetRoundedCorners: 5
D/DecorView(27057): semSetRoundedCorners: 5
D/DecorView(27057): semSetRoundedCorners: 5
D/DecorView(27057): semSetRoundedCorners: 5
D/DecorView(27057): semSetRoundedCorners: 5
...
```

Everything works well on individual screens.
I would be grateful for ideas on how to avoid this. And can this be done at all?

**Comment by aevzp** [comment]
Hi everyone.

After several attempts I came to the result
https://github.com/aevzp/flutter-gamepad

The solution is not perfect and carries a certain amount of limitations.

So, I am creating a native EditText field that I am focusing on it. This field I transfer to Flutter using Hybrid Composition. Using 'eventSink.success' I send buttons and axis events to Flutter. Everything seems to work, but there are performance issues. Such as:

**In my app I'm using WebRTC for video stream which returns the Texture. And If I display the video and add native EditText field for listening gamepad Controller, Texture widget is stuck and image does't update. As a solution is start animate something in Flutter. But I "feel" some lag especially on very weak devices.** 

If I turn on GPU debugging, I noticed some load spikes when I start to actively use the gamepad stick. As I understand I faces with a lot of calls 'eventSink.success' when interacting with gamepad.

For example is graph from Android 9:
![IMG_7058](https://user-images.githubusercontent.com/52816901/112129481-21b68400-8bd0-11eb-99bf-279e93e15ac7.jpg)

This is Android 10 (Green columns is interacting with gamepad stick):
![IMG_7059](https://user-images.githubusercontent.com/52816901/112129517-28dd9200-8bd0-11eb-913c-7dc47c1eefc8.jpg)

----

For now I can't find some good solution for listening gamepad stick except mine

Perhaps in the future it is planned to add an API with the ability to conveniently listen to gamepad sticks?
And what about freezing when I display `Texture` widget and Hybrid Composition? Any ideas?
Thank you.
